server/src/main/java/io/crate/auth/JWTAuthenticationMethod.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.auth;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.security.PublicKey;
+import java.security.interfaces.RSAPublicKey;
+import java.util.Locale;
+import java.util.function.Function;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import com.auth0.jwk.Jwk;
+import com.auth0.jwk.JwkException;
+import com.auth0.jwk.JwkProvider;
+import com.auth0.jwk.JwkProviderBuilder;
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.JWTVerifier;
+import com.auth0.jwt.algorithms.Algorithm;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.auth0.jwt.interfaces.Verification;
+
+import io.crate.protocols.postgres.ConnectionProperties;
+import io.crate.role.Role;
+import io.crate.role.Roles;
+
+public class JWTAuthenticationMethod implements AuthenticationMethod {
+
+    public static final String NAME = "jwt";
+
+    private final Roles roles;
+
+    private final Function<String, JwkProvider> urlToJwkProvider;
+
+
+    public JWTAuthenticationMethod(Roles roles, Function<String, JwkProvider> urlToJwkProvider) {
+        this.roles = roles;
+        this.urlToJwkProvider = urlToJwkProvider;
+    }
+
+    @Override
+    public @Nullable Role authenticate(Credentials credentials, ConnectionProperties connProperties) {
+        var username = credentials.username();
+        assert username != null : "User name must be resolved before authentication attempt";
+        var decodedJWT = credentials.decodedToken();
+        assert decodedJWT != null : "Token must be not null on jwt auth";
+
+        Role user = roles.findUser(username);
+        if (user == null) {
+            throw new RuntimeException("jwt authentication failed for user \"" + username + "\"");
```

Comment:
Is this correct? shouldn't we throw a dedicated authentication error, so we get a correct http error code? (or does this take place elsewhere)?

Reply:
it's aligned with `PasswordAuthenticationMethod`

Reply:
thx, is it maybe something we should take care of in general? (not in the context of this PR)

Reply:
not sure - now it's 500 error code and if we change exception/error code it might be a breaking change?
at least Runtime in auth worked so far for PwdAuth, so I aligned behavior with it.

Reply:
I'll bring it up for discussion, yep it's a breaking change so it's a + point to leave it as is, but let's see.